fdbserver/RestoreMaster.actor.cpp

```diff
@@ -0,0 +1,1339 @@
+					if ( !(param.length > 0  &&  param.offset >= 0 && param.offset < self->files[curFileIndex].fileSize) ) {
+						printf("[ERROR] param: length:%ld offset:%ld fileSize:%ld for %ldth filename:%s\n",
+								param.length, param.offset, self->files[curFileIndex].fileSize, curFileIndex,
+								self->files[curFileIndex].fileName.c_str());
+					}
+					ASSERT( param.length > 0 );
+					ASSERT( param.offset >= 0 );
+					ASSERT( param.offset < self->files[curFileIndex].fileSize );
+					self->files[curFileIndex].cursor = self->files[curFileIndex].cursor +  param.length;
+
+					RestoreCommandEnum cmdType = RestoreCommandEnum::Assign_Loader_Range_File;
+					if (self->files[curFileIndex].isRange) {
+						cmdType = RestoreCommandEnum::Assign_Loader_Range_File;
+						self->cmdID.setPhase(RestoreCommandEnum::Assign_Loader_Range_File);
+					} else {
+						cmdType = RestoreCommandEnum::Assign_Loader_Log_File;
+						self->cmdID.setPhase(RestoreCommandEnum::Assign_Loader_Log_File);
+					}
+
+					if ( (phaseType == RestoreCommandEnum::Assign_Loader_Log_File && self->files[curFileIndex].isRange) 
+						|| (phaseType == RestoreCommandEnum::Assign_Loader_Range_File && !self->files[curFileIndex].isRange) ) {
+						self->files[curFileIndex].cursor = 0;
+						curFileIndex++;
+						curOffset = 0;
+					} else { // load the type of file in the phaseType
+						self->cmdID.nextCmd();
+						printf("[CMD] Loading fileIndex:%ld fileInfo:%s loadingParam:%s on node %s\n",
+							curFileIndex, self->files[curFileIndex].toString().c_str(), 
+							param.toString().c_str(), loaderID.toString().c_str()); // VERY USEFUL INFO
+						printf("[INFO] Node:%s CMDUID:%s cmdType:%d isRange:%d loaderNode:%s\n", self->describeNode().c_str(), self->cmdID.toString().c_str(),
+								(int) cmdType, (int) self->files[curFileIndex].isRange, loaderID.toString().c_str());
+						if (self->files[curFileIndex].isRange) {
+							cmdReplies.push_back( loaderInterf.loadRangeFile.getReply(RestoreLoadFileRequest(self->cmdID, param)) );
+						} else {
+							cmdReplies.push_back( loaderInterf.loadLogFile.getReply(RestoreLoadFileRequest(self->cmdID, param)) );
+						}
+						curOffset += param.length;
+
+						// Reach the end of the file
+						if ( param.length + param.offset >= self->files[curFileIndex].fileSize ) {
+							curFileIndex++;
+							curOffset = 0;
+						}
+						
+						// if (param.length <= loadSizeB) { // Reach the end of the file
+						// 	ASSERT( self->files[curFileIndex].cursor == self->files[curFileIndex].fileSize );
+						// 	curFileIndex++;
+						// }
+					}
+					
+					if ( curFileIndex >= self->files.size() ) {
+						allLoadReqsSent = true;
+						break;
+					}
+					//++loadingCmdIndex; // Replaced by cmdUID
+				}
+
+				printf("[INFO] Wait for %ld loaders to accept the cmd Assign_Loader_File\n", cmdReplies.size());
+
+				// Question: How to set reps to different value based on cmdReplies.empty()?
```

Comment:
I'm not sure I understand the question, but between this and the TODO below, I'm going to guess that what you are looking for is:

maintain a vector<Future<Reply>> with size == # of loaders
wait( getAny( vector ) )
and then on each loop iteration, find any elements in the vector that are `isReady()`, and then give them a new loading request.

This way, you can keep a continuous stream of requests going to each loader, rather than sending one batch to each loader, and having to wait until the last one finishes to send out more loading requests.